### PR TITLE
Add well_segment.shift_logs

### PR DESCRIPTION
**Pull request by cdtn**

### CODE REVIEW

File: petroflow/src/well_segment.py [hunk 1]
```diff
@@ -1616,3 +1616,31 @@ def equalize_histogram(self, src=None, dst=None, channels='last'):
                 img = cv2.equalizeHist(img)
             setattr(self, _dst, img)
         return self
+
+    def shift_logs(self, mnemonics=None, max_period=3):
+        """Shift every logs column from `mnemonics` on random step, sampled from
+        discrete uniform distribution in range(-max_period, +max_period).
+        Parameters
```
Comment (Dimonovez) [inline petroflow/src/well_segment.py]: Add empty line.

File: petroflow/src/well_segment.py [hunk 2]
```diff
@@ -1616,3 +1616,31 @@ def equalize_histogram(self, src=None, dst=None, channels='last'):
...
+        ----------
+        mnemonics : None or str or list of str
+            - If `None`, shift all logs columns.
+            - If `str`, shift single column from logs with `mnemonics` name.
+            - If `list`, shift all logs columnns with names in `mnemonics`.
+            Defaults to `None`.
+        max_period : int
+            Max possible period absolute value.
+        Returns
```
Comment (Dimonovez) [inline petroflow/src/well_segment.py]: Add empty line.

File: petroflow/src/well_segment.py [hunk 3]
```diff
@@ -1616,3 +1616,31 @@ def equalize_histogram(self, src=None, dst=None, channels='last'):
                 img = cv2.equalizeHist(img)
             setattr(self, _dst, img)
         return self
+
+    def shift_logs(self, mnemonics=None, max_period=3):
+        """Shift every logs column from `mnemonics` on random step, sampled from
+        discrete uniform distribution in range(-max_period, +max_period).
```
Comment (Dimonovez) [inline petroflow/src/well_segment.py]: Maybe add more information about newly introduced missing values.

File: petroflow/src/well_segment.py [hunk 4]
```diff
@@ -1616,3 +1616,35 @@ def equalize_histogram(self, src=None, dst=None, channels='last'):
                 img = cv2.equalizeHist(img)
             setattr(self, _dst, img)
         return self
+
+    def shift_logs(self, mnemonics=None, max_period=3):
+        """Shift every logs column from `mnemonics` on random step sampled from
+        discrete uniform distribution in range(-max_period, +max_period). All
```
Comment (alexandrkuvaev) [inline petroflow/src/well_segment.py]:
```suggestion
        discrete uniform distribution in range (-max_period, +max_period). All
```

File: petroflow/src/well_segment.py [hunk 5]
```diff
@@ -1616,3 +1616,35 @@ def equalize_histogram(self, src=None, dst=None, channels='last'):
...
+        max_period : int, optional
+            Max possible period absolute value. Default is 3.
+
+        Returns
+        -------
+        self : AbstractWellSegment or a child class
+            Self with shifted logs columns.
+        """
+        mnemonics = self.logs.columns if mnemonics is None else to_list(mnemonics)
+        periods = np.random.randint(-max_period, max_period+1, len(mnemonics))
```
Comment (alexandrkuvaev) [inline petroflow/src/well_segment.py]:
```suggestion
        periods = np.random.randint(-max_period, max_period + 1, len(mnemonics))
```

File: petroflow/src/well_segment.py [hunk 6]
```diff
@@ -1616,3 +1616,35 @@ def equalize_histogram(self, src=None, dst=None, channels='last'):
...
+        for index, column in enumerate(mnemonics):
+            period = periods[index]
+            if period >= 0:
+                fill_value = self.logs[column].iloc[0]
+            else:
+                fill_value = self.logs[column].iloc[-1]
```
Comment (alexandrkuvaev) [inline petroflow/src/well_segment.py]: You can calculate an index of `fill_value` using ternary if in one line and then perform `iloc`.

File: petroflow/src/well_segment.py [hunk 7]
```diff
@@ -1616,3 +1616,35 @@ def equalize_histogram(self, src=None, dst=None, channels='last'):
...
+
+        Returns
+        -------
+        self : AbstractWellSegment or a child class
+            Self with shifted logs columns.
+        """
+        mnemonics = self.logs.columns if mnemonics is None else to_list(mnemonics)
+        periods = np.random.randint(-max_period, max_period+1, len(mnemonics))
+        for index, column in enumerate(mnemonics):
+            period = periods[index]
```
Comment (alexandrkuvaev) [inline petroflow/src/well_segment.py]: Just `zip` `mnemonics` and `periods`.

File: petroflow/src/well_segment.py [hunk 8]
```diff
@@ -1616,3 +1616,35 @@ def equalize_histogram(self, src=None, dst=None, channels='last'):
                 img = cv2.equalizeHist(img)
             setattr(self, _dst, img)
         return self
+
+    def shift_logs(self, mnemonics=None, max_period=3):
+        """Shift every logs column from `mnemonics` on random step sampled from
+        discrete uniform distribution in range(-max_period, +max_period). All
+        new resulting empty positions are filled with first/last columnn value
```
Comment (alexandrkuvaev) [inline petroflow/src/well_segment.py]:
```suggestion
        new resulting empty positions are filled with first/last column value
```

File: petroflow/src/well_segment.py [hunk 9]
```diff
@@ -1616,3 +1616,35 @@ def equalize_histogram(self, src=None, dst=None, channels='last'):
                 img = cv2.equalizeHist(img)
             setattr(self, _dst, img)
         return self
+
+    def shift_logs(self, mnemonics=None, max_period=3):
+        """Shift every logs column from `mnemonics` on random step sampled from
```
Comment (alexandrkuvaev) [inline petroflow/src/well_segment.py]:
```suggestion
        """Shift every `logs` column from `mnemonics` by a random step sampled from
```

File: petroflow/src/well_segment.py (hunk 4 above)
Comment (Dimonovez) [reply petroflow/src/well_segment.py]:
```suggestion
        discrete uniform distribution in range (-`max_period`,  `max_period`). All
```

File: petroflow/src/well_segment.py [hunk 10]
```diff
@@ -1616,3 +1616,33 @@ def equalize_histogram(self, src=None, dst=None, channels='last'):
...
+    def shift_logs(self, max_period, mnemonics=None):
+        """Shift every `logs` column from `mnemonics` by a random step sampled
+        from discrete uniform distribution in [-`max_period`, `max_period`].
+        All new resulting empty positions are filled with first/last
+        column value depending on shift direction.
+
+        Parameters
+        ----------
+        max_period : float
+            Max possible shift period length.
```
Comment (Dimonovez) [inline petroflow/src/well_segment.py]:
```suggestion
            Max possible shift period in meters.
```

**Pull request merged**